src/StreamJsonRpc.Tests/JsonRpcWithFatalExceptionsTests.cs

```diff
@@ -0,0 +1,304 @@
+        Assert.NotNull(exception.StackTrace);
+        Assert.Equal(exceptionMessage, this.serverRpc.FaultException.Message);
+        Assert.Equal(1, this.serverRpc.IsFatalExceptionCount);
+
+        // Assert that the JsonRpc and MessageHandler objects are disposed after exception
+        Assert.True(((IDisposableObservable)this.clientRpc).IsDisposed);
+        Assert.True(((IDisposableObservable)this.serverRpc).IsDisposed);
+        Assert.True(((DisposingMessageHandler)this.clientRpc.MessageHandler).IsDisposed);
+    }
+
+    [Fact]
+    public async Task CloseStreamOnAsyncYieldAndThrowException()
+    {
+        var exceptionMessage = "Exception from CloseStreamOnAsyncYieldAndThrowException";
+        Exception exception = await Assert.ThrowsAnyAsync<TaskCanceledException>(() => this.clientRpc.InvokeAsync<string>(nameof(Server.AsyncMethodThatThrowsAfterYield), exceptionMessage));
+        Assert.NotNull(exception.StackTrace);
+        Assert.Equal(exceptionMessage, this.serverRpc.FaultException.Message);
+        Assert.Equal(2, this.serverRpc.IsFatalExceptionCount);
+
+        // Assert that the JsonRpc and MessageHandler objects are disposed after exception
+        Assert.True(((IDisposableObservable)this.clientRpc).IsDisposed);
+        Assert.True(((IDisposableObservable)this.serverRpc).IsDisposed);
+        Assert.True(((DisposingMessageHandler)this.clientRpc.MessageHandler).IsDisposed);
+    }
+
+    [Fact]
+    public async Task CloseStreamOnAsyncThrowExceptionandYield()
+    {
+        var exceptionMessage = "Exception from CloseStreamOnAsyncThrowExceptionAndYield";
+        Exception exception = await Assert.ThrowsAnyAsync<TaskCanceledException>(() => this.clientRpc.InvokeAsync<string>(nameof(Server.AsyncMethodThatThrowsBeforeYield), exceptionMessage));
+        Assert.NotNull(exception.StackTrace);
+        Assert.Equal(exceptionMessage, this.serverRpc.FaultException.Message);
+        Assert.Equal(2, this.serverRpc.IsFatalExceptionCount);
+
+        // Assert that the JsonRpc and MessageHandler objects are disposed after exception
+        Assert.True(((IDisposableObservable)this.clientRpc).IsDisposed);
+        Assert.True(((IDisposableObservable)this.serverRpc).IsDisposed);
+        Assert.True(((DisposingMessageHandler)this.clientRpc.MessageHandler).IsDisposed);
+    }
+
+    [Fact]
+    public async Task CloseStreamOnAsyncMethodException()
+    {
+        var exceptionMessage = "Exception from CloseStreamOnAsyncMethodException";
+        await Assert.ThrowsAsync<TaskCanceledException>(() => this.clientRpc.InvokeAsync(nameof(Server.MethodThatThrowsAsync), exceptionMessage));
+        Assert.Equal(exceptionMessage, this.serverRpc.FaultException.Message);
+        Assert.Equal(2, this.serverRpc.IsFatalExceptionCount);
```

Comment:
I'd like to chat to better understand why this is `2` instead of just `1`.

Reply:
When I rethrew the exception we encountered it twice: once within the task handler and then once again when it was thrown from the task handler. I've changed it to close the streams within the task handler so there's no need to see the exception again.